lib/ruleHelper.js

```diff
@@ -76,15 +77,77 @@ RuleHelper.prototype = {
         } else if (expression.type === "CallExpression") {
             allowed = this.isAllowedCallExpression(expression.callee, escapeObject.methods || VALID_UNWRAPPERS);
         } else if (expression.type === "BinaryExpression") {
-            allowed = ((this.allowedExpression(expression.left, escapeObject))
-            && (this.allowedExpression(expression.right, escapeObject)));
+            allowed = ((this.allowedExpression(expression.left, escapeObject, details))
+            && (this.allowedExpression(expression.right, escapeObject, details)));
         } else if (expression.type === "TSAsExpression") {
             // TSAsExpressions contain the raw javascript value in 'expression'
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (expression.type === "TypeCastExpression") {
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (Array.isArray(expression)) {
-            allowed = expression.every((e) => this.allowedExpression(e, escapeObject));
+            allowed = expression.every((e) => this.allowedExpression(e, escapeObject, details));
+        } else if (expression.type === "Identifier") {
+            // find declared variables and see which are literals
+            const scope = this.context.getScope(expression);
+            const variableInfo = scope.set.get(expression.name);
+
+            // let's see if this variable is allowed, by tracing it back through the code
+            if (variableInfo) {
+                // look if the var was defined as allowable
+                let definedAsAllowed = false;
+                for (const def of variableInfo.defs) {
```

Comment:
Nit (really optional, more of a discussion point): If we early return on `variableInfo.defs.length == 0`, we could rewrite this using `every` ([1]), but I think we should do it only if it does really improve the readibility otherwise the for is just fine.

I think that the resulting code using every should look like more or less as:
```js
      if (!variableInfo || variableInfo.defs.length === 0) {
        // details.message = `Variable ${expression.name} considered unsafe: variable initialization not found`;
        return false;
      }
 
      // let's see if this variable is allowed, by tracing it back through the code
      const definedAsAllowed = variableInfo.defs.every((def) => {
          if (def.node.type !== "VariableDeclarator") {
              // TODO: maybe report this as unsupported syntax to file a bug for (as suggested in another review comment)
              return false;
          }
          if ((def.kind !== "let") && (def.kind !== "const")) {
              const {line, column} = varInitAs.loc.start;
              details.message = `Identifiers declared using '${def.kind}' are never allowed. `+
                `Declare '${expression.name}' using "let" or "const" to allow it when set to a safe value`;
              return false;
          }
          // the `init` property carries the right-hand side of the variable definition:
          const varInitAs = def.node.init;
          if (!this.allowedExpression(varInitAs, escapeObject, details)) {
            // if one variable definition is considered unsafe, all are.
            // NB: order of definition is unclear. See issue #168.
            if (!details.message) {
              const {line, column} = varInitAs.loc.start;
              details.message = `Variable '${expression.name}' initialized with unsafe value at ${line}:${column}`;
            }
            return false;
          }

          return true;
      });

      if (!definedAsAllowed) {
        return false;
      }
      ... // then something similar for variableInfo.references
```

(same comment is also valid for the for loop over `variableInfo.references` from a few lines below)

-----

[1]: `every` always return true for an empty array and so to fully replace this with every we need to be sure we are not calling it on an empy array

Reply:
I've never used `every` before and wow, it's been there for a while.

It seems a useful replacement of the "for of", especially since we want "every" item to validate according to a specific function. But I somehow feel a bit uneasy about this. Maybe because of habit? :|

Do you think the for-of is too complicated?

Reply:
> Do you think the for-of is too complicated?

@mozfreddyb no, I don't think that it is too much complicated, `every` "may" make it a bit more explicit that we want that every entry of the array to be allowed but in the end the for-of is fine as it is.

Let's not worry about this right now. Feel free to leave the for-of and resolve this comment.